File: Mediconnet-Backend/Services/NotificationService.cs
using Microsoft.EntityFrameworkCore;
using Microsoft.AspNetCore.SignalR;
using Mediconnet_Backend.Core.Interfaces.Services;
using Mediconnet_Backend.Core.Entities;
using Mediconnet_Backend.Data;
using Mediconnet_Backend.Hubs;

namespace Mediconnet_Backend.Services;

/// <summary>
/// Service centralisé de gestion des notifications
/// </summary>
public class NotificationService : INotificationService
{
    private readonly ApplicationDbContext _context;
    private readonly IHubContext<NotificationHub> _hubContext;
    private readonly ILogger<NotificationService> _logger;

    public NotificationService(
        ApplicationDbContext context,
        IHubContext<NotificationHub> hubContext,
        ILogger<NotificationService> logger)
    {
        _context = context;
        _hubContext = hubContext;
        _logger = logger;
    }

    // ==================== CRÉATION ====================

    public async Task<NotificationDto> CreateAsync(CreateNotificationRequest request)
    {
        var notification = new Notification
        {
            IdUser = request.IdUser,
            Type = request.Type,
            Titre = request.Titre,
            Message = request.Message,
            Lien = request.Lien,
            Icone = request.Icone ?? GetDefaultIcon(request.Type),
            Priorite = request.Priorite,
            DateExpiration = request.DateExpiration,
            Metadata = request.Metadata
        };

        _context.Notifications.Add(notification);
        await _context.SaveChangesAsync();

        var dto = MapToDto(notification);

        if (request.SendRealTime)
        {
            await SendRealTimeNotificationAsync(request.IdUser, dto);
        }

        _logger.LogInformation("Notification créée: {Id} pour user {UserId}", notification.IdNotification, request.IdUser);

        return dto;
    }

    public async Task<List<NotificationDto>> CreateBulkAsync(CreateBulkNotificationRequest request)
    {
        var notifications = request.UserIds.Select(userId => new Notification
        {
            IdUser = userId,
            Type = request.Type,
            Titre = request.Titre,
            Message = request.Message,
            Lien = request.Lien,
            Icone = request.Icone ?? GetDefaultIcon(request.Type),
            Priorite = request.Priorite,
            DateExpiration = request.DateExpiration,
            Metadata = request.Metadata
        }).ToList();

        _context.Notifications.AddRange(notifications);
        await _context.SaveChangesAsync();

        var dtos = notifications.Select(MapToDto).ToList();

        if (request.SendRealTime)
        {
            foreach (var (notification, index) in notifications.Select((n, i) => (n, i)))
            {
                await SendRealTimeNotificationAsync(notification.IdUser, dtos[index]);
            }
        }

        _logger.LogInformation("Notifications bulk créées: {Count} notifications", notifications.Count);

        return dtos;
    }

    public async Task<int> CreateForRoleAsync(string role, CreateNotificationRequest request)
    {
        var userIds = await _context.Utilisateurs
            .Where(u => u.Role == role)
            .Select(u => u.IdUser)
            .ToListAsync();

        if (!userIds.Any()) return 0;

        var bulkRequest = new CreateBulkNotificationRequest
        {
            UserIds = userIds,
            Type = request.Type,
            Titre = request.Titre,
            Message = request.Message,
            Lien = request.Lien,
            Icone = request.Icone,
            Priorite = request.Priorite,
            DateExpiration = request.DateExpiration,
            Metadata = request.Metadata,
            SendRealTime = request.SendRealTime
        };

        var created = await CreateBulkAsync(bulkRequest);
        return created.Count;
    }

    // ==================== LECTURE ====================

    public async Task<NotificationListResult> GetUserNotificationsAsync(int userId, NotificationFilter? filter = null)
    {
        filter ??= new NotificationFilter();

        var query = _context.Notifications
            .Where(n => n.IdUser == userId && !n.Supprime)
            .AsQueryable();

        // Filtres
        if (!string.IsNullOrEmpty(filter.Type))
            query = query.Where(n => n.Type == filter.Type);

        if (filter.Lu.HasValue)
            query = query.Where(n => n.Lu == filter.Lu.Value);

        if (!string.IsNullOrEmpty(filter.Priorite))
            query = query.Where(n => n.Priorite == filter.Priorite);

        if (filter.DateDebut.HasValue)
            query = query.Where(n => n.DateCreation >= filter.DateDebut.Value);

        if (filter.DateFin.HasValue)
            query = query.Where(n => n.DateCreation <= filter.DateFin.Value);

        // Exclure les expirées
        query = query.Where(n => n.DateExpiration == null || n.DateExpiration > DateTime.UtcNow);

        var totalCount = await query.CountAsync();
        var unreadCount = await query.CountAsync(n => !n.Lu);

        var notifications = await query
            .OrderByDescending(n => n.DateCreation)
            .Skip((filter.Page - 1) * filter.PageSize)
            .Take(filter.PageSize)
            .ToListAsync();

        return new NotificationListResult
        {
            Notifications = notifications.Select(MapToDto).ToList(),
            TotalCount = totalCount,
            UnreadCount = unreadCount,
            Page = filter.Page,
            PageSize = filter.PageSize
        };
    }

    public async Task<NotificationDto?> GetByIdAsync(int notificationId)
    {
        var notification = await _context.Notifications.FindAsync(notificationId);
        return notification != null ? MapToDto(notification) : null;
    }

    public async Task<int> GetUnreadCountAsync(int userId)
    {
        return await _context.Notifications
            .CountAsync(n => n.IdUser == userId && !n.Lu && !n.Supprime &&
                (n.DateExpiration == null || n.DateExpiration > DateTime.UtcNow));
    }

    // ==================== MISE À JOUR ====================

    public async Task<bool> MarkAsReadAsync(int notificationId, int userId)
    {
        var notification = await _context.Notifications
            .FirstOrDefaultAsync(n => n.IdNotification == notificationId && n.IdUser == userId);

        if (notification == null) return false;

        notification.Lu = true;
        notification.DateLecture = DateTime.UtcNow;
        await _context.SaveChangesAsync();

        // Envoyer mise à jour du compteur
        await SendUnreadCountUpdateAsync(userId);

        return true;
    }

    public async Task<int> MarkAllAsReadAsync(int userId)
    {
        var notifications = await _context.Notifications
            .Where(n => n.IdUser == userId && !n.Lu && !n.Supprime)
            .ToListAsync();

        foreach (var notification in notifications)
        {
            notification.Lu = true;
            notification.DateLecture = DateTime.UtcNow;
        }

        await _context.SaveChangesAsync();

        // Envoyer mise à jour du compteur
        await SendUnreadCountUpdateAsync(userId);

        return notifications.Count;
    }

    public async Task<int> MarkMultipleAsReadAsync(int userId, List<int> notificationIds)
    {
        var notifications = await _context.Notifications
            .Where(n => notificationIds.Contains(n.IdNotification) && n.IdUser == userId && !n.Lu)
            .ToListAsync();

        foreach (var notification in notifications)
        {
            notification.Lu = true;
            notification.DateLecture = DateTime.UtcNow;
        }

        await _context.SaveChangesAsync();

        // Envoyer mise à jour du compteur
        await SendUnreadCountUpdateAsync(userId);

        return notifications.Count;
    }

    // ==================== SUPPRESSION ====================

    public async Task<bool> DeleteAsync(int notificationId, int userId)
    {
        var notification = await _context.Notifications
            .FirstOrDefaultAsync(n => n.IdNotification == notificationId && n.IdUser == userId);

        if (notification == null) return false;

        notification.Supprime = true;
        await _context.SaveChangesAsync();

        // Envoyer mise à jour du compteur si non lue
        if (!notification.Lu)
        {
            await SendUnreadCountUpdateAsync(userId);
        }

        return true;
    }

    public async Task<int> DeleteAllReadAsync(int userId)
    {
        var notifications = await _context.Notifications
            .Where(n => n.IdUser == userId && n.Lu && !n.Supprime)
            .ToListAsync();

        foreach (var notification in notifications)
        {
            notification.Supprime = true;
        }

        await _context.SaveChangesAsync();
        return notifications.Count;
    }

    public async Task<int> CleanupExpiredAsync()
    {
        var expired = await _context.Notifications
            .Where(n => n.DateExpiration != null && n.DateExpiration < DateTime.UtcNow && !n.Supprime)
            .ToListAsync();

        foreach (var notification in expired)
        {
            notification.Supprime = true;
        }

        await _context.SaveChangesAsync();

        _logger.LogInformation("Nettoyage: {Count} notifications expirées supprimées", expired.Count);

        return expired.Count;
    }

    // ==================== TEMPS RÉEL ====================

    public async Task SendRealTimeNotificationAsync(int userId, NotificationDto notification)
    {
        try
        {
            await _hubContext.Clients.Group($"user_{userId}")
                .SendAsync("NewNotification", notification);

            // Envoyer aussi le nouveau compteur
            var unreadCount = await GetUnreadCountAsync(userId);
            await _hubContext.Clients.Group($"user_{userId}")
                .SendAsync("UnreadCountUpdate", unreadCount);
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Erreur envoi notification temps réel à user {UserId}", userId);
        }
    }

    public async Task SendRealTimeNotificationToGroupAsync(string group, NotificationDto notification)
    {
        try
        {
            await _hubContext.Clients.Group(group)
                .SendAsync("NewNotification", notification);
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Erreur envoi notification temps réel au groupe {Group}", group);
        }
    }

    private async Task SendUnreadCountUpdateAsync(int userId)
    {
        try
        {
            var unreadCount = await GetUnreadCountAsync(userId);
            await _hubContext.Clients.Group($"user_{userId}")
                .SendAsync("UnreadCountUpdate", unreadCount);
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Erreur envoi mise à jour compteur à user {UserId}", userId);
        }
    }

    // ==================== HELPERS ====================

    private static NotificationDto MapToDto(Notification n) => new()
    {
        IdNotification = n.IdNotification,
        IdUser = n.IdUser,
        Type = n.Type,
        Titre = n.Titre,
        Message = n.Message,
        Lien = n.Lien,
        Icone = n.Icone,
        Priorite = n.Priorite,
        Lu = n.Lu,
        DateLecture = n.DateLecture,
        DateCreation = n.DateCreation,
        Metadata = n.Metadata
    };

    private static string GetDefaultIcon(string type) => type switch
    {
        NotificationType.RendezVous => "calendar",
        NotificationType.Facture => "credit-card",
        NotificationType.Consultation => "stethoscope",
        NotificationType.Alerte => "alert-triangle",
        NotificationType.AlerteMedicale => "heart-pulse",
        NotificationType.Stock => "package",
        NotificationType.Systeme => "settings",
        NotificationType.Message => "message-circle",
        NotificationType.Rappel => "bell",
        NotificationType.Validation => "check-circle",
        _ => "bell"
    };
}

File: Mediconnet-Frontend/src/app/services/notification.service.ts
import { Injectable, OnDestroy, Inject, forwardRef } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { BehaviorSubject, Observable, Subject } from 'rxjs';
import { takeUntil, tap, catchError } from 'rxjs/operators';
import * as signalR from '@microsoft/signalr';
import { environment } from '../../environments/environment';
import { NotificationSoundService } from './notification-sound.service';

// ==================== INTERFACES ====================

export interface Notification {
  idNotification: number;
  idUser: number;
  type: string;
  titre: string;
  message: string;
  lien?: string;
  icone?: string;
  priorite: string;
  lu: boolean;
  dateLecture?: Date;
  dateCreation: Date;
  metadata?: string;
  tempsEcoule?: string;
}

export interface NotificationFilter {
  type?: string;
  lu?: boolean;
  priorite?: string;
  dateDebut?: Date;
  dateFin?: Date;
  page?: number;
  pageSize?: number;
}

export interface NotificationListResult {
  notifications: Notification[];
  totalCount: number;
  unreadCount: number;
  page: number;
  pageSize: number;
  totalPages: number;
}

export interface CreateNotificationRequest {
  idUser: number;
  type: string;
  titre: string;
  message: string;
  lien?: string;
  icone?: string;
  priorite?: string;
  dateExpiration?: Date;
  metadata?: string;
  sendRealTime?: boolean;
}

// ==================== SERVICE ====================

@Injectable({
  providedIn: 'root'
})
export class NotificationService implements OnDestroy {
  private readonly apiUrl = `${environment.apiUrl}/notification`;
  private readonly hubUrl = `${environment.apiUrl.replace('/api', '')}/hubs/notifications`;
  
  private hubConnection: signalR.HubConnection | null = null;
  private destroy$ = new Subject<void>();
  
  // État réactif
  private notificationsSubject = new BehaviorSubject<Notification[]>([]);
  private unreadCountSubject = new BehaviorSubject<number>(0);
  private loadingSubject = new BehaviorSubject<boolean>(false);
  private connectedSubject = new BehaviorSubject<boolean>(false);

  // Observables publics
  public notifications$ = this.notificationsSubject.asObservable();
  public unreadCount$ = this.unreadCountSubject.asObservable();
  public loading$ = this.loadingSubject.asObservable();
  public connected$ = this.connectedSubject.asObservable();

  constructor(
    private http: HttpClient,
    private soundService: NotificationSoundService
  ) {}

  // ==================== CONNEXION SIGNALR ====================

  /**
   * Initialise la connexion SignalR pour les notifications temps réel
   */
  async startConnection(token: string): Promise<void> {
    if (this.hubConnection?.state === signalR.HubConnectionState.Connected) {
      return;
    }

    this.hubConnection = new signalR.HubConnectionBuilder()
      .withUrl(this.hubUrl, {
        accessTokenFactory: () => token
      })
      .withAutomaticReconnect([0, 2000, 5000, 10000, 30000])
      .configureLogging(signalR.LogLevel.Warning)
      .build();

    // Gestionnaires d'événements SignalR
    this.hubConnection.on('NewNotification', (notification: Notification) => {
      this.handleNewNotification(notification);
    });

    this.hubConnection.on('UnreadCountUpdate', (count: number) => {
      this.unreadCountSubject.next(count);
    });

    this.hubConnection.on('NotificationMarkedAsRead', (notificationId: number) => {
      this.updateNotificationReadStatus(notificationId, true);
    });

    this.hubConnection.onreconnecting(() => {
      console.log('Reconnexion aux notifications...');
      this.connectedSubject.next(false);
    });

    this.hubConnection.onreconnected(() => {
      console.log('Reconnecté aux notifications');
      this.connectedSubject.next(true);
      this.loadUnreadCount();
    });

    this.hubConnection.onclose(() => {
      console.log('Déconnecté des notifications');
      this.connectedSubject.next(false);
    });

    try {
      await this.hubConnection.start();
      console.log('Connecté aux notifications temps réel');
      this.connectedSubject.next(true);
      await this.loadUnreadCount();
    } catch (err) {
      console.error('Erreur connexion SignalR notifications:', err);
      this.connectedSubject.next(false);
    }
  }

  /**
   * Arrête la connexion SignalR
   */
  async stopConnection(): Promise<void> {
    if (this.hubConnection) {
      await this.hubConnection.stop();
      this.hubConnection = null;
      this.connectedSubject.next(false);
    }
  }

  // ==================== API CALLS ====================

  /**
   * Récupère les notifications de l'utilisateur
   */
  getNotifications(filter?: NotificationFilter): Observable<NotificationListResult> {
    let params = new HttpParams();
    
    if (filter) {
      if (filter.type) params = params.set('type', filter.type);
      if (filter.lu !== undefined) params = params.set('lu', filter.lu.toString());
      if (filter.priorite) params = params.set('priorite', filter.priorite);
      if (filter.page) params = params.set('page', filter.page.toString());
      if (filter.pageSize) params = params.set('pageSize', filter.pageSize.toString());
    }

    this.loadingSubject.next(true);
    
    return this.http.get<NotificationListResult>(this.apiUrl, { params }).pipe(
      tap(result => {
        this.notificationsSubject.next(result.notifications);
        this.unreadCountSubject.next(result.unreadCount);
        this.loadingSubject.next(false);
      }),
      catchError(err => {
        this.loadingSubject.next(false);
        throw err;
      })
    );
  }

  /**
   * Charge le nombre de notifications non lues
   */
  loadUnreadCount(): Promise<void> {
    return new Promise((resolve, reject) => {
      this.http.get<{ count: number }>(`${this.apiUrl}/unread-count`).subscribe({
        next: (result) => {
          this.unreadCountSubject.next(result.count);
          resolve();
        },
        error: (err) => {
          console.error('Erreur chargement compteur notifications:', err);
          reject(err);
        }
      });
    });
  }

  /**
   * Récupère une notification par ID
   */
  getById(id: number): Observable<Notification> {
    return this.http.get<Notification>(`${this.apiUrl}/${id}`);
  }

  /**
   * Marque une notification comme lue
   */
  markAsRead(id: number): Observable<any> {
    return this.http.patch(`${this.apiUrl}/${id}/read`, {}).pipe(
      tap(() => {
        this.updateNotificationReadStatus(id, true);
        const currentCount = this.unreadCountSubject.value;
        if (currentCount > 0) {
          this.unreadCountSubject.next(currentCount - 1);
        }
      })
    );
  }

  /**
   * Marque toutes les notifications comme lues
   */
  markAllAsRead(): Observable<{ message: string; count: number }> {
    return this.http.patch<{ message: string; count: number }>(`${this.apiUrl}/read-all`, {}).pipe(
      tap(() => {
        const notifications = this.notificationsSubject.value.map(n => ({ ...n, lu: true }));
        this.notificationsSubject.next(notifications);
        this.unreadCountSubject.next(0);
      })
    );
  }

  /**
   * Marque plusieurs notifications comme lues
   */
  markMultipleAsRead(ids: number[]): Observable<{ message: string; count: number }> {
    return this.http.patch<{ message: string; count: number }>(`${this.apiUrl}/read-multiple`, { ids }).pipe(
      tap(() => {
        const notifications = this.notificationsSubject.value.map(n => 
          ids.includes(n.idNotification) ? { ...n, lu: true } : n
        );
        this.notificationsSubject.next(notifications);
      })
    );
  }

  /**
   * Supprime une notification
   */
  delete(id: number): Observable<any> {
    return this.http.delete(`${this.apiUrl}/${id}`).pipe(
      tap(() => {
        const notifications = this.notificationsSubject.value.filter(n => n.idNotification !== id);
        this.notificationsSubject.next(notifications);
      })
    );
  }

  /**
   * Supprime toutes les notifications lues
   */
  deleteAllRead(): Observable<{ message: string; count: number }> {
    return this.http.delete<{ message: string; count: number }>(`${this.apiUrl}/read`).pipe(
      tap(() => {
        const notifications = this.notificationsSubject.value.filter(n => !n.lu);
        this.notificationsSubject.next(notifications);
      })
    );
  }

  // ==================== HELPERS ====================

  private handleNewNotification(notification: Notification): void {
    const currentNotifications = this.notificationsSubject.value;
    this.notificationsSubject.next([notification, ...currentNotifications]);
    
    // Jouer le son de notification selon la priorité
    this.soundService.playNotificationSound(notification.priorite);
    
    // Afficher une notification système si disponible
    this.showBrowserNotification(notification);
  }

  private updateNotificationReadStatus(id: number, lu: boolean): void {
    const notifications = this.notificationsSubject.value.map(n => 
      n.idNotification === id ? { ...n, lu } : n
    );
    this.notificationsSubject.next(notifications);
  }

  /**
   * Affiche une notification navigateur si autorisé
   */
  private async showBrowserNotification(notification: Notification): Promise<void> {
    // Vérifier si les notifications navigateur sont supportées et autorisées
    if (!('Notification' in window)) return;
    
    if (Notification.permission === 'granted') {
      try {
        new Notification(notification.titre, {
          body: notification.message,
          icon: '/assets/images/logo-icon.png',
          tag: `notif-${notification.idNotification}`,
          silent: true // Le son est géré par notre service
        });
      } catch (e) {
        // Service Worker non disponible ou erreur
      }
    }
  }

  /**
   * Demande la permission pour les notifications navigateur
   */
  async requestBrowserNotificationPermission(): Promise<boolean> {
    if (!('Notification' in window)) return false;
    
    if (Notification.permission === 'granted') return true;
    if (Notification.permission === 'denied') return false;
    
    const permission = await Notification.requestPermission();
    return permission === 'granted';
  }

  /**
   * Calcule le temps écoulé depuis la création
   */
  getTempsEcoule(dateCreation: Date | string): string {
    const date = new Date(dateCreation);
    const now = new Date();
    const diffMs = now.getTime() - date.getTime();
    const diffMins = Math.floor(diffMs / 60000);
    const diffHours = Math.floor(diffMs / 3600000);
    const diffDays = Math.floor(diffMs / 86400000);

    if (diffMins < 1) return "À l'instant";
    if (diffMins < 60) return `Il y a ${diffMins} min`;
    if (diffHours < 24) return `Il y a ${diffHours}h`;
    if (diffDays < 7) return `Il y a ${diffDays}j`;
    return date.toLocaleDateString('fr-FR');
  }

  /**
   * Retourne l'icône par défaut selon le type
   */
  getIconForType(type: string): string {
    const icons: Record<string, string> = {
      'rdv': 'calendar',
      'facture': 'credit-card',
      'consultation': 'stethoscope',
      'alerte': 'alert-triangle',
      'alerte_medicale': 'heart-pulse',
      'stock': 'package',
      'systeme': 'settings',
      'message': 'message-circle',
      'rappel': 'bell',
      'validation': 'check-circle'
    };
    return icons[type] || 'bell';
  }

  /**
   * Retourne la couleur selon la priorité
   */
  getColorForPriority(priorite: string): string {
    const colors: Record<string, string> = {
      'basse': '#64748b',
      'normale': '#3b82f6',
      'haute': '#f59e0b',
      'urgente': '#ef4444'
    };
    return colors[priorite] || colors['normale'];
  }

  ngOnDestroy(): void {
    this.destroy$.next();
    this.destroy$.complete();
    this.stopConnection();
  }
}
